typegen/src/typegen/settings/substitutes.rs

```diff
@@ -18,9 +18,15 @@ fn error(span: Span, kind: TypeSubstitutionErrorKind) -> TypeSubstitutionError {
 /// to figure out when to swap said type with some provided substitute.
 #[derive(Debug, Clone)]
 pub struct TypeSubstitutes {
-    substitutes: HashMap<syn::Path, Substitute>,
+    substitutes: HashMap<PathSegments, Substitute>,
 }
 
+/// We use this `Vec<String>` as a common denominator, since we need a consistent key for both
+/// `syn::TypePath` and `scale_info::ty::path::Path` types.
+///
+/// Can be obtained from a [`syn::Path`] via the [`path_segments`] function.
+pub type PathSegments = Vec<String>;
```

Comment:
Does it make any sense to hide this behind a newtype so that only valid `PathSegments` can be constructed?

Eg below we do `.expect("PathSegments should be syn::PathSegment compatible")`, but we can have any arbitrary `Vec<String>` here?

Reply:
Hmm, I am not a big fan of `PathSegments(Vec<String>)` because it caused us a lot of allocations converting to it (cloning the Vec and every String in it, every time we want to check if a type should be a substitute). I thought about having a borrowed type like `PathSegments<'a>(Cow<'a, str>)` but I think a simple validation on insert might be enough. What do you think?